### ISSUE TITLE
Erroneous consumption value 16777.215 (gas)

### ISSUE THREAD

**Issue by matroderick** [n3rgy/data#1]
Occasionally, SMEST2 gas meters have been seen to report an anomalously high consumption value, reporting 16777.215 (m3) within single 30min periods. This is a result of the meter failing to capture the consumption value correctly and reporting the hex value 0xFFFFFF (which is 167777215 in decimal). This is a "feature" of the SMETS specification, the specification uses the 0xFFFFFF value within the consumption field as an error code (which we believe isn't best practice but is what is defined within the specification). 

n3rgy are considering the best approach to improve this situation, in the mean time please update your code to detect and correct these values appropriately.

UPDATE
We have also discovered some meter manufactures are using a similar mechanism (0xFFF??) to report other error codes which are not defined within the specification. Therefore we have suspended our plans to intercept these values until further evidence is available.

**Comment by richplane** [comment]
Would be good to have a rough ETA for this fix so we know whether it's worth adding code to catch it at our end - it's blocking our data import at the moment for a couple of meters.

**Comment by matroderick** [comment]
hi @richplane, the fix is in the backlog for the next sprint (so release mid September). If possible, I'd suggest you implement your work around as there's almost zero chance of 167777.215 being returned as a valid consumption value and therefore your work around should not have any enduring implication on your service (once our change is released).

Once updated our service will return the following in the event the meter returns an invalid value (as an example)

{
    "value": 0.0,
    "timestamp": "2019-12-30 23:30",
    "status": "INVALID"
}

**Comment by richplane** [comment]
Just wanted to add that we're also seeing a high value of 188250.978 coming through from gas meters, which looks like it's the FFFFFF error value multiplied by a m3 > kWh conversion factor of ~11.22. I guess this will be caught by the same fix?

**Comment by matroderick** [comment]
Yes, the consumption value with the smart meter is in thousands of m3 so a 0xFFFFFF error value is divided by 1,000 to get to m3 then the kWh conversion factor can be applied. As a platform we do not mandate what conversion factor is used by our users.

**Comment by snigehere** [comment]
Did anything ever happen solve this issue?  I have am still seeing this data in my export from n3rgy.com
Multiple instances last month again

timestamp (UTC) energyConsumption (m3)

06/07/2022 00:30 16777.215
11/07/2022 00:30 16777.215
16/07/2022 00:30 16777.215
17/07/2022 00:30 16777.215
18/07/2022 00:30 16777.215
22/07/2022 00:30 16777.215
25/07/2022 00:30 16777.215
26/07/2022 00:30 16777.215
30/07/2022 00:30 16777.215
05/08/2022 00:30 16777.215

**Comment by matroderick** [comment]
Sorry, this is a "feature" of the meters when they fail to record the real consumption value.

**Comment by snigehere** [comment]
Thanks.. yes.. I recognise its a "feature" and this is an error message.. my question was really about getting them filtered on the data side .. eg fix it once rather than every user having to filter them out.

**Comment by pdcastro** [comment]
> _which we believe isn't best practice but is what is defined within the specification_

A fair point, but I imagine that the spec envisaged some relatively low-level software interpreting a raw byte sequence equal to `0xFFFFFF`, rather than a CSV spreadsheet displaying `16777.215` as consumption data to an end(ish) user.

In a CSV spreadsheet, better options than `16777.215` might include:

* A blank cell, representing the fact that _“no actual energy consumption data is available for this cell”._
* The `"NaN"` string ([Not a Number](https://en.wikipedia.org/wiki/NaN)), commonly displayed in place of an unviable number (e.g. the result of multiplying infinity by zero).
* A string such as `"Error 0xFFFFFF"`, which would allow the error code information to be preserved.

I think a blank cell might be best as it is probably the least disruptive change in terms of backwards compatibility.

**Comment by pdcastro** [comment]
By the way, my previous comment was from the point of view of an end user of n3rgy’s Consumer Portal, [Download My Data](https://data.n3rgy.com/consumer/download-data) web tool. I saw `16777.215` in a CSV spreadsheet that I had downloaded, googled it and came across this GitHub issue. A “blank CSV cell” doesn’t quite apply to a JSON API, but even in a JSON API, simply returning value `16777.215` doesn’t sound like the best approach.
